crates/router/src/core/refunds_v2.rs

```diff
@@ -0,0 +1,798 @@
+                            refund_status: Some(enums::RefundStatus::Failure),
+                            refund_error_message: Some(
+                                errors::ConnectorError::NotImplemented(message.to_owned())
+                                    .to_string(),
+                            ),
+                            refund_error_code: Some("NOT_IMPLEMENTED".to_string()),
+                            updated_by: storage_scheme.to_string(),
+                            connector_refund_id: None,
+                            processor_refund_data: None,
+                            unified_code: None,
+                            unified_message: None,
+                        })
+                    }
+                    errors::ConnectorError::NotSupported { message, connector } => {
+                        Some(storage::RefundUpdate::ErrorUpdate {
+                            refund_status: Some(enums::RefundStatus::Failure),
+                            refund_error_message: Some(format!(
+                                "{message} is not supported by {connector}"
+                            )),
+                            refund_error_code: Some("NOT_SUPPORTED".to_string()),
+                            updated_by: storage_scheme.to_string(),
+                            connector_refund_id: None,
+                            processor_refund_data: None,
+                            unified_code: None,
+                            unified_message: None,
+                        })
+                    }
+                    _ => None,
+                });
+        // Update the refund status as failure if connector_error is NotImplemented
+        if let Some(refund_error_update) = option_refund_error_update {
+            state
+                .store
+                .update_refund(
+                    refund.to_owned(),
+                    refund_error_update,
+                    merchant_account.storage_scheme,
+                )
+                .await
+                .to_not_found_response(errors::ApiErrorResponse::InternalServerError)
+                .attach_printable_lazy(|| {
+                    format!(
+                        "Failed while updating refund: refund_id: {}",
+                        refund.id.get_string_repr()
+                    )
+                })?;
+        }
+        let mut refund_router_data_res = router_data_res.to_refund_failed_response()?;
+        // Initiating Integrity check
+        let integrity_result = check_refund_integrity(
+            &refund_router_data_res.request,
+            &refund_router_data_res.response,
+        );
+        refund_router_data_res.integrity_check = integrity_result;
+        refund_router_data_res
+    } else {
+        router_data
+    };
+
+    let refund_update = match router_data_res.response {
+        Err(err) => {
+            let option_gsm = helpers::get_gsm_record(
```

Comment:
Have one helper which gives option_gsm in the core flow, move these branching into a function.